### PR TITLE
CLN: @doc - base.py & indexing.py

### PR DESCRIPTION
**Pull request by HH-MWB**
- [x] working on #31942 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by HH-MWB** [comment]
Replacing all docstring decorator at once will be too hard to accomplish. I would prefer to do it step-by-step. Clean a few / one file in a PR, and do things iteratively. 

Would you mind help me review the current changes? At the mean time, I will keep working on next part.

**Comment by WillAyd** [comment]
@HH-MWB can you post screenshots of the affected docstrings?

**Review by simonjayhawkins — COMMENTED** [review]
Thanks @HH-MWB generally lgtm, however a few questions.

**Comment by HH-MWB** [comment]
> @HH-MWB can you post screenshots of the affected docstrings?

@WillAyd  No problem. Should I use `help()` or `.__doc__`?

File: pandas/core/indexing.py
Comment (HH-MWB) [reply pandas/core/indexing.py]: Thanks @simonjayhawkins! Good catch. I would need to figure out a way to avoid this f-string. 

I put f-string here is because the doc here use `{` and `}`, and we want to use this docstring latter in other functions.

**Comment by WillAyd** [comment]
You can build the doc strings and just post a screenshot of the HTML

File: pandas/core/indexes/datetimelike.py
Comment (HH-MWB) [reply pandas/core/indexes/datetimelike.py]: All three make good sense to me, and I think `Datetime-like Index` may looks better. What do you think?

When I was updating this, I focused more on using `doc` decorator to replace the `Appender` and `Substitution`. So that I try to avoid changes of other parts. Adding `klass` is because there is a place holder for that and we might not leave it empty. I am not sure about your second question. BTW, how might I know if it published by Sphinx?

**Comment by HH-MWB** [comment]
> You can build the doc strings and just post a screenshot of the HTML

Get! Will do, thanks!

File: pandas/core/indexing.py
Comment (HH-MWB) [reply pandas/core/indexing.py]: > Thanks @simonjayhawkins! Good catch. I would need to figure out a way to avoid this f-string.

I made [a small tweak in doc decorator](https://github.com/pandas-dev/pandas/pull/31970/commits/6cd80820e938a15dc5b1e76bb2f6bf079c47b0f4#diff-e41db70762f6c5b076f9b551651d4132), and it will fix the issue here.

File: pandas/core/indexes/datetimelike.py
Comment (simonjayhawkins) [reply pandas/core/indexes/datetimelike.py]: > BTW, how might I know if it published by Sphinx?

sorry that I wasn't clear. I just meant the published docs at https://dev.pandas.io/docs/index.html

Comment (simonjayhawkins) [reply pandas/core/indexes/datetimelike.py]: > When I was updating this, I focused more on using `doc` decorator to replace the `Appender` and `Substitution`. So that I try to avoid changes of other parts.

sure

File: pandas/core/indexing.py
Comment (simonjayhawkins) [reply pandas/core/indexing.py]: > I made [a small tweak in doc decorator](https://github.com/pandas-dev/pandas/pull/31970/commits/6cd80820e938a15dc5b1e76bb2f6bf079c47b0f4#diff-e41db70762f6c5b076f9b551651d4132), and it will fix the issue here.

ok

**Review by simonjayhawkins — APPROVED** [review]
thanks @HH-MWB lgtm. would rather https://github.com/pandas-dev/pandas/pull/31970#discussion_r379617289 was addressed here but could be a follow-up.

would it be feasible to inherit the docstring template from the super class instead of explicitly specifying it? It would address my concern for Categorical array.

**Comment by HH-MWB** [comment]
> thanks @HH-MWB lgtm. would rather [#31970 (comment)](https://github.com/pandas-dev/pandas/pull/31970#discussion_r379617289) was addressed here but could be a follow-up.
> 
> would it be feasible to inherit the docstring template from the super class instead of explicitly specifying it? It would address my concern for Categorical array.

@simonjayhawkins Sorry for the late reply. 

First, I agree with you that **inherit the docstring from the super class is very intuitive**, and also could saves us a lot of time. Then, when I dive deeper for this case, I found that the original docstring is different with it's super class. This difference confused me and I am trying to figure it out what cause this difference and how can we solve it.


### Inherit Docstring
I think it is a very good idea to inherit the docstring. Maybe this should not be a default behavior but at least we should be able to do inheritance easily. 

A potential quick solution for this is to create another decorator, for example, `doc_inhert`. I would expect it to bring the first non-empty docstring for it's super class. 

For example, `B.f` should have the docstring inhert from `A.f`:
```python
class A:
    def f(self):
        """func doc"""

class B(A):
    @doc_inhert
    def f(self):
        pass
```

Another example, `C.f` should have the docstring inhert from `A.f`, because `B.f` is empty:
```python
class A:
    def f(self):
        """func doc"""

class B(A):
    def f(self):
        pass

class C(B):
    @doc_inhert
    def f(self):
        pass
```

I am still looking to see if there are other better solutions. For example, one thing in my mind will be create decorator for the class, and do docstring inhert for all methods belong to that class.

Would you mind if I create a new issue about docstring inheritance and propose my idea there? I am not 100% sure, but I think we already have some issues about this. 

Open for suggestions and advices. What should I do next?

### Case Here

As what I mentioned above, the original docstring is different from it's super class's docstring. Should I keep it as the original one or switch to be same as the super class one?

Here is the original docstring:
```
Find indices where elements should be inserted to maintain order.

Find the indices into a sorted {klass} `self` such that, if the
corresponding elements in `value` were inserted before the indices,
the order of `self` would be preserved.

.. note::

    The {klass} *must* be monotonically sorted, otherwise
    wrong locations will likely be returned. Pandas does *not*
    check this for you.

Parameters
----------
value : array_like
    Values to insert into `self`.
side : {{'left', 'right'}}, optional
    If 'left', the index of the first suitable location found is given.
    If 'right', return the last such index.  If there is no suitable
    index, return either 0 or N (where N is the length of `self`).
sorter : 1-D array_like, optional
    Optional array of integer indices that sort `self` into ascending
    order. They are typically the result of ``np.argsort``.

Returns
-------
int or array of int
    A scalar or array of insertion points with the
    same shape as `value`.

    .. versionchanged:: 0.24.0
        If `value` is a scalar, an int is now always returned.
        Previously, scalar inputs returned an 1-item array for
        :class:`Series` and :class:`Categorical`.

See Also
--------
sort_values
numpy.searchsorted

Notes
-----
Binary search is used to find the required insertion points.

Examples
--------
>>> x = pd.Series([1, 2, 3])
>>> x
0    1
1    2
2    3
dtype: int64

>>> x.searchsorted(4)
3

>>> x.searchsorted([0, 4])
array([0, 3])

>>> x.searchsorted([1, 3], side='left')
array([0, 2])

>>> x.searchsorted([1, 3], side='right')
array([1, 3])

>>> x = pd.Categorical(['apple', 'bread', 'bread',
                        'cheese', 'milk'], ordered=True)
[apple, bread, bread, cheese, milk]
Categories (4, object): [apple < bread < cheese < milk]

>>> x.searchsorted('bread')
1

>>> x.searchsorted(['bread'], side='right')
array([3])

If the values are not monotonically sorted, wrong locations
may be returned:

>>> x = pd.Series([2, 1, 3])
>>> x.searchsorted(1)
0  # wrong result, correct would be 1
```

Here is the docstring from it's super class:
```
Find indices where elements should be inserted to maintain order.

.. versionadded:: 0.24.0

Find the indices into a sorted array `self` (a) such that, if the
corresponding elements in `value` were inserted before the indices,
the order of `self` would be preserved.

Assuming that `self` is sorted:

======  ================================
`side`  returned index `i` satisfies
======  ================================
left    ``self[i-1] < value <= self[i]``
right   ``self[i-1] <= value < self[i]``
======  ================================

Parameters
----------
value : array_like
    Values to insert into `self`.
side : {'left', 'right'}, optional
    If 'left', the index of the first suitable location found is given.
    If 'right', return the last such index.  If there is no suitable
    index, return either 0 or N (where N is the length of `self`).
sorter : 1-D array_like, optional
    Optional array of integer indices that sort array a into ascending
    order. They are typically the result of argsort.

Returns
-------
array of ints
    Array of insertion points with the same shape as `value`.

See Also
--------
numpy.searchsorted : Similar method from NumPy.
```

**Comment by HH-MWB** [comment]
> You can build the doc strings and just post a screenshot of the HTML

Hi @WillAyd, I might mis-understand what you said. 

I create HTML via `python make.py html`, but because all files changed is under pandas.core and pandas.util, the html doc might not cover those things.
> The pandas.core, pandas.compat, and pandas.util top-level modules are PRIVATE. Stable functionality in such modules is not guaranteed.

**Comment by simonjayhawkins** [comment]
> Would you mind if I create a new issue about docstring inheritance and propose my idea there? I am not 100% sure, but I think we already have some issues about this.
> 
> Open for suggestions and advices. What should I do next?

Thanks for the detail. It looks like we should definitely have a separate issue/PR for this discussion.

**Comment by simonjayhawkins** [comment]
> I create HTML via `python make.py html`, but because all files changed is under pandas.core and pandas.util, the html doc might not cover those things.

am I correct in thinking the only published docstring affected by this PR is https://dev.pandas.io/docs/reference/api/pandas.Series.searchsorted.html? If so the screenshot of that should be OK.

**Comment by HH-MWB** [comment]
> am I correct in thinking the only published docstring affected by this PR is https://dev.pandas.io/docs/reference/api/pandas.Series.searchsorted.html? If so the screenshot of that should be OK.

Thanks for your help @simonjayhawkins! Here is the screen shots. 

Screen Shot 2020-02-16 at 2 40 57 AM

Screen Shot 2020-02-16 at 2 42 47 AM

**Comment by simonjayhawkins** [comment]
The screenshots are from locally generated documentation?

**Comment by HH-MWB** [comment]
> The screenshots are from locally generated documentation?

Sorry for the confusion, but yes, it is locally generated doc.

File: pandas/core/arrays/categorical.py
Comment (WillAyd) [reply pandas/core/arrays/categorical.py]: I agree with this comment here - this seems strange to use `IndexOpsMixin` since it isn't part of this classes hierarchy. Any reason not to address this?

File: pandas/core/indexes/datetimelike.py
Comment (WillAyd) [inline pandas/core/indexes/datetimelike.py]: Same comment - this is very confusing

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/datetimelike.py
Comment (HH-MWB) [reply pandas/core/indexes/datetimelike.py]: Hi @WillAyd, I agree with you and @ simonjayhawkins here. It might be confusing. However, the original docstring is not extended from the base class. It seems like the original code obscures the problem because it does not explicitly indicate the source of the docstring. I try to keep it as it is but use `@doc`.

It looks like we all agree this docstring template will confuse other developers, but do you feel needed to fix this issue in this PR? If so, what will be your suggestion? One option that comes in my mind will be using the docstring from the base class, and modify them to fit in this case. I was trying to avoid that because it will change the original docstring relations, and I am not sure if we did this on purpose. 

Just to be clarified, I am very willing to make the additional change to solve this confusion. I just don't know what will be the best way of doing that. One more thing, I have a [comment](https://github.com/pandas-dev/pandas/pull/31970#issuecomment-586631930) related to this. You might also be interested.

File: pandas/util/_decorators.py
Comment (datapythonista) [inline pandas/util/_decorators.py]: I don't understand this, do you mind explaining why is this needed?

**Review by datapythonista — COMMENTED** [review]
Changes look good, I think we can change where the `searchsorted` docstring is injected later, and keep things simple here.

Just a question about the curly brackets replacement.

File: pandas/util/_decorators.py
Comment (HH-MWB) [reply pandas/util/_decorators.py]: Hi, a short quick answer: 
This change solves the error when we want to reference the docstring from function which not using doc decorator.

You must already know this, but just want to clarify this again so that helps others catch up:
When we format string in Python, we use `{` and `}` as a place holder, and use the arguments to replace the place holder. For example, `'Hello {name}!'.format(name='Tom')` will become `'Hello Tom!'`. However, if we want to show `{` or `}` from a formatted string, we have to put `{{` and `}}`. For example, `'Hello {{name}}!'.format(name='Tom')` will become `'Hello {Tom}!'`. 

Back to our case: 
In the docstring, `{` and `}` are very commonly used to represent set. The old implementation will just take the original docstring, and then `{` and `}` will be considered as a place holder. This will cause an error. We can fix that by switching `{` and `}` to `{{` and `}}` in the doc decorator. Then, it will be shown as `{` and `}`.

Please take the changes in [this file](https://github.com/pandas-dev/pandas/blob/a1b0e0bfd1a1c8bf65b69f5acf298390da57dd02/pandas/core/indexing.py) as an example:
The docstring of `IndexingMixin.iloc` has `{` and `}`, these changes will allow us to do something like [this](https://github.com/pandas-dev/pandas/blob/a1b0e0bfd1a1c8bf65b69f5acf298390da57dd02/pandas/core/indexing.py#L1346):
```python
@doc(IndexingMixin.iloc)
class _iLocIndexer(_LocationIndexer):
```

I am not sure if I explain the reason clear enough, please feel free to come with follow up questions.

Comment (datapythonista) [reply pandas/util/_decorators.py]: Sorry, I still don't get it. I see these cases:

**Not using `@doc`**:
- Enumeration: `{'a', 'b'}`
- Variable: Doesn't apply

**Using `@doc`**:
- Enumeration: `{{'a', 'b'}}`
- Variable: `{name}`

I guess what this solves is taking a docstring from a non-decorated function, and use it with the decorator, converting it from the first case to the second. But this feels a bit hacky, and I'm not sure if it would make more sense to apply `.format()` only when it makes sense instead. I'm a bit worried on making this too complex and tricky, and having unexpected side-cases.

And if we need to keep this implementation, we need to add comments so a reader can understand what's going on, and what these `if` cases mean.

Comment (HH-MWB) [reply pandas/util/_decorators.py]: > I guess what this solves is taking a docstring from a non-decorated function, and use it with the decorator, converting it from the first case to the second.

Yes, exactly! Thanks for helping me explain this. That is want I try to say.

> ... I'm not sure if it would make more sense to apply .format() only when it makes sense instead.

That make sense to me. I would like to give it a try. 

> I'm a bit worried on making this too complex and tricky, and having unexpected side-cases.

I agree with you here. I would like to try the idea above first, and a backup solution might be using string template to avoid this case.

Comment (HH-MWB) [reply pandas/util/_decorators.py]: >  ... I'm not sure if it would make more sense to apply .format() only when it makes sense instead.

Hi @datapythonista, I have modified the `@doc` to apply `.format()` only when it is coming from decorated function or docstring.

It seems there is some issue with the CI checks. I will keep working on this, but also very welcome to get some help here.

File: pandas/util/_decorators.py
Comment (simonjayhawkins) [inline pandas/util/_decorators.py]: can you confirm that these `# type: ignore` stem from the fact that we are adding attributes to the wrapper function?

I guess this was discussed originally, but would using a class for the doc decorator help?

Comment (HH-MWB) [reply pandas/util/_decorators.py]: Good point here! Most `# type: ignore` is because we are adding attributes to function. 

The only one not for this reason is `else dedent(arg.__doc__)  # type: ignore` (line 288). This one is because theoretically, `arg.__doc__` can be `str` or `None`. However, we have checked `arg.__doc__` is not `None` before putting it in, so that we are safe here.

Back to the most `# type: ignore` cases. I strongly agree with you that we could consider solving this issue without using `# type: ignore`, because this looks like an abuse. I would like to try putting it under a class. Thanks for this advice!

**Comment by HH-MWB** [comment]
Hello @simonjayhawkins. It seems very hard to remove all `# type: ignore`, because we have to bind a variable to the function, instead of the decorator. 

However, we can reduce using `# type: ignore` from 6 times to 3 times. 

Here is my proposal by using class:
```python
class doc:
    """
    A decorator take docstring templates, concatenate them and perform string
    substitution on it.

    This decorator will add a variable "_docs" to the wrapped function to keep
    track the original docstring template for potential usage. If it should be
    consider as a template, it will be saved as a string. Otherwise, it will be
    saved as callable, and later user __doc__ and dedent to get docstring.

    Parameters
    ----------
    *args : str or callable
        The string / docstring / docstring template to be appended in order
        after default docstring under function.
    **kwags : str
        The string which would be used to format docstring template.
    """

    def __init__(self, *args: Union[str, Callable], **kwargs: str) -> None:
        self._docs: List[Union[str, Callable]] = []
        for arg in args:
            if hasattr(arg, "_docs"):
                self._docs.extend(arg._docs)  # type: ignore
            elif isinstance(arg, str) or arg.__doc__:
                self._docs.append(arg)

        self.parameters = kwargs

    def __call__(self, func: F) -> F:
        if func.__doc__:
            self._docs.insert(0, dedent(func.__doc__))

        func.__doc__ = "".join(
            [
                arg.format(**self.parameters)
                if isinstance(arg, str)
                else dedent(arg.__doc__)  # type: ignore
                for arg in self._docs
            ]
        )

        func._docs = self._docs  # type: ignore

        return func
```

I think we can use same idea to reduce `# type: ignore` from using function.

File: pandas/core/indexes/datetimelike.py
Comment (WillAyd) [reply pandas/core/indexes/datetimelike.py]: Is there a more suitable location for the docstring then? Importing the IndexOpsMixin here is strange

Comment (datapythonista) [reply pandas/core/indexes/datetimelike.py]: Personally I think this should be addressed in a follow up. It's here in `IndexOpsMixin` because that's where the `_shared_docs` was. I think this PR is already too complex to make the change here. And if we change `_shared_docs` before merging this, the conflict here will be quite annoying to fix. Does it make sense?

File: pandas/util/_decorators.py
Comment (simonjayhawkins) [reply pandas/util/_decorators.py]: > The only one not for this reason is `else dedent(arg.__doc__) # type: ignore` (line 288). This one is because theoretically, `arg.__doc__` can be `str` or `None`. However, we have checked `arg.__doc__` is not `None` before putting it in, so that we are safe here.

i _think_ ok to remove
```
$ mypy pandas --warn-unused-ignores
pandas\util\_decorators.py:288: error: unused 'type: ignore' comment
```

**Comment by simonjayhawkins** [comment]
> Hello @simonjayhawkins. It seems very hard to remove all `# type: ignore` because we might have to bind a variable to the function instead of the decorator.
> 
> For use case of `@doc(f, ...)`, we should be able to get original templates from function `f`. If there are any other way might work, please let me know, and I will be more than happy to give it a try.
> 
> **However, we can reduce using `# type: ignore` from 6 times to 3 times.**
> 
> Here is my proposal by using class:
> 
> ```python
> class doc:
>     """
>     A decorator take docstring templates, concatenate them and perform string
>     substitution on it.
> 
>     This decorator will add a variable "_docs" to the wrapped function to keep
>     track the original docstring template for potential usage. If it should be
>     consider as a template, it will be saved as a string. Otherwise, it will be
>     saved as callable, and later user __doc__ and dedent to get docstring.
> 
>     Parameters
>     ----------
>     *args : str or callable
>         The string / docstring / docstring template to be appended in order
>         after default docstring under function.
>     **kwags : str
>         The string which would be used to format docstring template.
>     """
> 
>     def __init__(self, *args: Union[str, Callable], **kwargs: str) -> None:
>         self._docs: List[Union[str, Callable]] = []
>         for arg in args:
>             if hasattr(arg, "_docs"):
>                 self._docs.extend(arg._docs)  # type: ignore
>             elif isinstance(arg, str) or arg.__doc__:
>                 self._docs.append(arg)
> 
>         self.parameters = kwargs
> 
>     def __call__(self, func: F) -> F:
>         if func.__doc__:
>             self._docs.insert(0, dedent(func.__doc__))
> 
>         func.__doc__ = "".join(
>             [
>                 arg.format(**self.parameters)
>                 if isinstance(arg, str)
>                 else dedent(arg.__doc__)  # type: ignore
>                 for arg in self._docs
>             ]
>         )
> 
>         func._docs = self._docs  # type: ignore
> 
>         return func
> ```
> 
> We can use same idea to reduce `# type: ignore` from using function:
> 
> ```python
> def doc(*args: Union[str, Callable], **kwargs: str) -> Callable[[F], F]:
>     """
>     A decorator take docstring templates, concatenate them and perform string
>     substitution on it.
> 
>     This decorator will add a variable "_docs" to the wrapped function to keep
>     track the original docstring template for potential usage. If it should be
>     consider as a template, it will be saved as a string. Otherwise, it will be
>     saved as callable, and later user __doc__ and dedent to get docstring.
> 
>     Parameters
>     ----------
>     *args : str or callable
>         The string / docstring / docstring template to be appended in order
>         after default docstring under function.
>     **kwags : str
>         The string which would be used to format docstring template.
>     """
> 
>     def decorator(func: F) -> F:
>         @wraps(func)
>         def wrapper(*args, **kwargs) -> Callable:
>             return func(*args, **kwargs)
> 
>         # collecting docstring and docstring templates
>         docs: List[Union[str, Callable]] = []
>         if func.__doc__:
>             docs.append(dedent(func.__doc__))
> 
>         for arg in args:
>             if hasattr(arg, "_docs"):
>                 docs.extend(arg._docs)  # type: ignore
>             elif isinstance(arg, str) or arg.__doc__:
>                 docs.append(arg)
> 
>         # formatting templates and concatenating docstring
>         wrapper.__doc__ = "".join(
>             [
>                 arg.format(**kwargs)
>                 if isinstance(arg, str)
>                 else dedent(arg.__doc__)  # type: ignore
>                 for arg in docs
>             ]
>         )
> 
>         # saving docstring templates list
>         wrapper._docs = docs  # type: ignore
> 
>         return cast(F, wrapper)
> 
>     return decorator
> ```
> 
> Open for feed backs here. Which one is better? Any other improvement suggestions please?

probably best to leave as a follow-up and have more detailed discussion and just keep the changes to the decorator in this PR limited to the changes needed for using a docstring from a method that is not decorated.

File: pandas/util/_decorators.py
Comment (HH-MWB) [reply pandas/util/_decorators.py]: @simonjayhawkins Good catch here! Thanks. [Fixing](https://github.com/pandas-dev/pandas/pull/31970/commits/7bf242e03bc10aeee4a7c6df515df23fbca23859) is here.

File: pandas/core/indexes/datetimelike.py
Comment (HH-MWB) [reply pandas/core/indexes/datetimelike.py]: > Is there a more suitable location for the docstring then? Importing the IndexOpsMixin here is strange

For me, I feel this situation might be caused by they are sharing the same interface, but don't have a common ancestor. If we have an interface declaration, I would say putting docstring there would be a good choice. However, Python as a dynamic programming language, we don't have to declare the interface.

Now, I don't have a good solution in mind, but I would like to look for it and see if we can find somewhere that makes more sense.

**Comment by HH-MWB** [comment]
> > Open for feed backs here. Which one is better? Any other improvement suggestions please?
> 
> probably best to leave as a follow-up and have more detailed discussion and just keep the changes to the decorator in this PR limited to the changes needed for using a docstring from a method that is not decorated.

Sounds very reasonable. I will follow-up latter.

File: pandas/util/_decorators.py
Comment (datapythonista) [inline pandas/util/_decorators.py]: I think the name `_doc_args` is misleading. In the context of this function makes sense, since it's the value of the `args` parameters of the `doc` decorator, but if I see a `_doc_args` attribute of for example `pandas.Series.head`, I don't think it really tells what's in it. Something like `_docstring_components` could be clearer?

Also, when can it contain a callable? Even when we have something like `@doc(pandas.Series.head)` we're extending it with the content templates of `pandas.Series.head`, not with the function.

Last thing, what about using a `docstring_components` variable to avoid all the `type: ignore`, and just set `wrapper._docstring_components = _docstring_components` at the end? Probably cleaner.

File: pandas/util/_decorators.py
Comment (datapythonista) [inline pandas/util/_decorators.py]: There is a trailing whitespace in this line breaking the CI.

**Review by datapythonista — COMMENTED** [review]
There is a trailing space and a mypy error, other than that looks good.

File: pandas/util/_decorators.py
Comment (HH-MWB) [reply pandas/util/_decorators.py]: @datapythonista Thanks for this suggestion, especially for the new variable name. I strongly agree with you here.

**Comment by HH-MWB** [comment]
> There is a trailing space and a mypy error, other than that looks good.

@datapythonista, @simonjayhawkins. Sorry to bother you. This [mypy error](https://github.com/pandas-dev/pandas/pull/31970/checks?check_run_id=492785846) makes me very confused.

In [this commit](https://github.com/pandas-dev/pandas/pull/31970/commits/7bf242e03bc10aeee4a7c6df515df23fbca23859), I removed `type: ignore` but everything works fine. However, now it shows the type error. I am confused why the original code can pass without error but the new code can't.

I know we can easily fix this by adding `type: ignore`, but I would like to know what case this problem and if there are a better way to solve that.

File: pandas/util/_decorators.py
Comment (datapythonista) [inline pandas/util/_decorators.py]: The decorator can be used like this:
```python
@doc
def foo():
    pass

@doc(foo)
def bar():
    pass
```
If that happens, in this line `arg` will be the function `foo`, and `arg.__doc__` will be None. When `dedent` is called with `None` as parameter we get:
```
>>> dedent(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mgarcia/miniconda3/lib/python3.7/textwrap.py", line 430, in dedent
    text = _whitespace_only_re.sub('', text)
TypeError: expected string or bytes-like object
```
And mypy is warning you about it. There are different ways of fixing this. May be this?
```suggestion
                arg.format(**kwargs) if isinstance(arg, str) else dedent(arg.__doc__ or "")
```

File: pandas/util/_decorators.py
Comment (datapythonista) [inline pandas/util/_decorators.py]: Probably not worth having this comment.

File: pandas/util/_decorators.py
Comment (HH-MWB) [reply pandas/util/_decorators.py]: That make sense.

In line 281 and 282, there are codes to make sure `arg.__doc__` won't be `None`.
```python
elif isinstance(arg, str) or arg.__doc__:
    docstring_components.append(arg)
```

In this case, maybe we can use `# type: ignore`?

Comment (datapythonista) [reply pandas/util/_decorators.py]: I think that code ensures `_docstring_components` is not set to `None` (in the parent). And the exception won't happen as far as `_docstring_components` is always set with the `@doc` decorator. But technically speaking there is no guarantee that something like this can be done:
```python
def foo():
    pass
foo._docstring_components = [None]

@doc(foo)
def bar():
    pass
```

So, my preference is an actual fix, like the one I proposed, than a typing `ignore`.

Comment (HH-MWB) [reply pandas/util/_decorators.py]: > I think that code ensures `_docstring_components` is not set to `None` (in the parent). And the exception won't happen as far as `_docstring_components` is always set with the `@doc` decorator. But technically speaking there is no guarantee that something like this can be done:
> 
> ```python
> def foo():
>     pass
> foo._docstring_components = [None]
> 
> @doc(foo)
> def bar():
>     pass
> ```
> 
> So, my preference is an actual fix, like the one I proposed, than a typing `ignore`.

That is a good point! Error fixed. Thanks!

File: pandas/util/_decorators.py
Comment (datapythonista) [inline pandas/util/_decorators.py]: Not sure if this could be a better fix for this one:
```suggestion
            if hasattr(arg, "_docstring_components") and isinstance(arg._docstring_components, list):
                docstring_components.extend(arg._docstring_components)
```

**Review by datapythonista — APPROVED** [review]
lgtm, just one idea that could make sense.

Thanks for the work on this @HH-MWB, it's a great improvement.

File: pandas/util/_decorators.py
Comment (HH-MWB) [reply pandas/util/_decorators.py]: Hi @datapythonista. I would consider this as two parts. One is to add type check for `_docstring_components` and one for remove `# type: ignore`.

#### For remove `# type: ignore` 
After I made the change, I still get the errors below and not sure if we can remove that.
```bash
pandas/util/_decorators.py:280: error: Item "str" of "Union[str, Callable[..., Any]]" has no attribute "_docstring_components"
pandas/util/_decorators.py:280: error: Item "function" of "Union[str, Callable[..., Any]]" has no attribute "_docstring_components"
```

*I guess my settings are not the same as checks here. I even the last passed build, I still get mypy errors on local. So, the change might fix it and just I don't know.*

#### For checking if `_docstring_components` is `list`
I think this is a very good point, we do as defensive programming. But I am a little bit curious about how far we should go to protect it. Also, since this function is internal use for pandas only, would it be better to let it failed? Then the developer would get notified there is a conflict immediately?

Just to be clarified, I would be happy to make this change if you (or anyone) feels needed. It is just I didn't see a clear improvement on this, because I see both pros and cons for this.

File: pandas/core/arrays/categorical.py
Comment (WillAyd) [inline pandas/core/arrays/categorical.py]: I believe I am in the minority on this view and don't want to be overly difficult, but can you refactor the `IndexOpsMixin` as a pre-cursor to this, or leave this module separate from the rest of changes (which look good btw)

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/categorical.py
Comment (HH-MWB) [reply pandas/core/arrays/categorical.py]: Yes, that is a good idea. I would convert this back to using `_shared_docs`. This PR is already too large, and I should really put them separately.

File: pandas/util/_decorators.py
Comment (datapythonista) [reply pandas/util/_decorators.py]: I'm happy to merge this as is, was wondering if the suggested change would fix the problem.

We can surely go back to it in the future if needed, but let's merge this first, since this already adds a lot of value, and we can start replacing `Appender` by `doc` at a larger scale after this.

**Review by datapythonista — APPROVED** [review]
Can you merge master. I guess the CI problems are unrelated.

**Comment by WillAyd** [comment]
Great thanks @HH-MWB ! If you can do a follow up to fix the IndexOpsMixin import would be very much appreciated!

**Comment by HH-MWB** [comment]
> Great thanks @HH-MWB ! If you can do a follow up to fix the IndexOpsMixin import would be very much appreciated!

Yes. I am thinking what is the best way of doing that. Definitely want to keep making contributing on this! Thanks.

**Pull request merged**
